### ISSUE TITLE
Request profile parameter from `application/vc`

### ISSUE THREAD

**Issue by OR13** [w3c/vc-data-model#1363]
For example `application/ld+json` already supports profile:

```
application/ld+json; profile="https://www.w3.org/ns/credentials"
```

`application/vc+...` could similarly support profile:

```
application/vc+ld+json; profile="https://market-vertical.example/ns/credentials"
```

See the related thread from the CG list:

- https://lists.w3.org/Archives/Public/public-credentials/2023Nov/0080.html

> So, if you're doing a VCDM v2.0 VC secured using Anoncreds, your VC
> will end up having this as the context (because DataIntegrityProof is
> included by default in VCDM v2.0):

> "`@context`": [
> "https://www.w3.org/ns/credentials/v2",
> "https://market-vertical.example/contexts/v1"
> ]

> and if you're doing a VCDM v1.1 VC secured using Anoncreds, your VC
> ends up w/ this context (because DataIntegrityProof IS NOT included by
> default in VCDM v1.0 or v1.1):

> "`@context`": [
>   "https://www.w3.org/2018/credentials/v1",
>   "https://w3id.org/security/data-integrity/v2",
>   "https://market-vertical.example/contexts/v1"
> ]

> and if Anoncreds needs to define new properties (which I'd very
> strongly advise against), your VC ends up w/ this context:

> "`@context`": [
>  "https://www.w3.org/2018/credentials/v1",
>  "https://anoncreds.example/contexts/v1",
>  "https://market-vertical.example/contexts/v1"
> ]

**Comment by OR13** [comment]
We have a related issue open regarding `vc-jose-cose` at https://github.com/w3c/vc-jose-cose/issues/184

I suppose there are also considerations regarding profile and data integrity, for example:

```
application/ld+json; profile="https://w3id.org/security/data-integrity"
```

^ The above would signal the use of data integrity as defined by https://github.com/perma-id/w3id.org which could be the same as or different from:

```
application/ld+json; profile="https://www.w3.org/ns/security/data-integrity"
``

The security namespace (`https://www.w3.org/ns/security`) only appears in data integrity afaik... https://w3c.github.io/vc-data-integrity/#example-capability-invocation-property-containing-two-verification-methods

`https://www.w3.org/ns/security` does not appear in the core data model specification.

**Comment by OR13** [comment]
I raised https://github.com/w3c/vc-jose-cose/pull/186 which removes part of this issue from `vc-jose-cose`, but it remains an issue in the core data model.

For example, a JOSE or COSE credential could have the following `typ` values now:

```
"typ":  "application/jwt",
...
"typ":  "application/sd-jwt",
...
"typ":  "application/cose",
...
"typ":  "application/jwt; profile=42", // assuming profile was registered.
...
"typ":  "application/cwt; profile=42", // assuming profile was registered.
```

By not mandating a specific value for `typ` we protect against ossification, which was raised as a concern here:

https://mailarchive.ietf.org/arch/msg/media-types/CFUhb5yGd4ix5oCjBf4LVmaPk3k/

We have the same issue with `cty` values / the media types requested to be registered in the core data model.

```
"cty":  "application/vc+ld+json",
...
"cty":  "application/vc+ld+json; profile=v2?",
...
"cty":  "application/vc2+ld+json; profile=v2?",
...
```

Does `application/vc+ld+json` work with v1 or only v2 ?

See the related registration request from activity streams:

https://www.w3.org/TR/activitystreams-core/#media-type

They opted to use just `+json` and a profile parameter... maybe we should do the same?

**Comment by OR13** [comment]
I suggest:

1. marking the registrations at risk
2. copy something like the text we have in did core.
3. add text providing guidance on parameters
4. add warnings about parameters

**Comment by OR13** [comment]
recently raised (related) : 

- https://github.com/w3c/vc-data-model/issues/1364

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-11-28-vcwg-special#section1-4) on 2023-11-28

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.4. Request profile parameter from `application/vc` (issue vc-data-model#1363)

_See github issue [vc-data-model#1363](https://github.com/w3c/vc-data-model/issues/1363)._




**Brent Zundel:** The current path we are on relies on media types with multiple suffixes.  
… The spec for multiple suffixes in IETF has been more contentious than anticipated.  
… And possibly a whole lot more work than anticipated.  
… This issue is another path to sidestep this.  

**Orie Steele:** This issue can be used with or without multiple suffixes.  
… Our current registrations require multiple suffixes, even though this isn't possible at present.  
… Even for application/vc the subtype can contain media type parameters.  
… Those parameters can express characteristics of the media type.  
… Specifying the text type is a common usage.  
… This tracks the intended use of the profile parameter.  
… We should provide guidance on the use of profile.  
… We could say not to use it.  
… We could say to use it in the same way as ld+json.  

**Brent Zundel:** Thank you for clarifying the relationship between the parameters.  
… We can register media types with multiple suffixes.  
… The current interpretation of a+b+c is that a is the prefix and b+c is the suffix.  

> *Orie Steele:* we can request registrations that rely on multiple suffixes, but they cannot be registered, until their interpretation is clear... it is a mistake to request registrations that we know cannot be approved.

**Ted Thibodeau Jr.:** The first thing after the slash isn't what matters.  
… The subtype is what's after the last + sign.  
… There are multiple layers of subtyping.  
… I don't know why the IETF did this.  

> *Orie Steele:* if you have questions regarding media types, you can ask the media type list. see [https://datatracker.ietf.org/doc/draft-ietf-mediaman-suffixes/](https://datatracker.ietf.org/doc/draft-ietf-mediaman-suffixes/).

**Ted Thibodeau Jr.:** You process as many of the subtypes that you understand.  
… This is complex and esoteric.  

> *Orie Steele:* +1 TallTed, its complicated : (.

**Ted Thibodeau Jr.:** The problem with multiple + signs is that the interpretation is undefined.  
… There may be many more drafts.  
… We can specify what our media types mean but those will only apply to people who understand our spec.  
… It's to let people who don't understand what's to the left of the +s to still work with what's to the right of the +s.  

> *Orie Steele:* here is the repo: [https://github.com/ietf-wg-mediaman/suffixes](https://github.com/ietf-wg-mediaman/suffixes) ... TallTed, eager to see your PR : ).

**Manu Sporny:** I am the editor of that spec.  
… There is only one remaining issue.  
… It's about whether to register all the things in between.  

> *Dave Longley:* for "application/vc+ld+json" processing is: "application", then "json", then "ld+json", then "vc+ld+json" ... another way to understand this subclass processing is: "application/square+rectangle+shape" ... you can understand any "application" content, then, if you want, more specifically any "shape", ... then any "rectangle" ... "square".

**Manu Sporny:** The chair of the WG wants to take it to last call.  
… I don't think the multiple structured suffixes draft is in a place that it's not going to be ratified soon.  
… Ted is right that we can just specify what our media types mean.  
… We don't need to change anything about the media types used in the spec.  
… The profile problem is problemantic.  
… Many people don't prrocess it.  

> *Orie Steele:* My recommendation would be to not depend on multiple suffixes, we raised a PR to remove the dependency from vc-jose-cose ... [https://github.com/w3c/vc-jose-cose/pull/186](https://github.com/w3c/vc-jose-cose/pull/186).

> *Orie Steele:* related registration for [https://www.iana.org/assignments/media-types/application/ld+json](https://www.iana.org/assignments/media-types/application/ld+json).

> *Orie Steele:* related specification Manu mentioned: [https://www.w3.org/TR/activitystreams-core/#media-type](https://www.w3.org/TR/activitystreams-core/#media-type).

**Manu Sporny:** You can have multiple different IETF registrations stomp over it.  
… I realize we used it in JSON-LD. I wish we hadn't.  
… We can tell people not to use it.  
… What we've learned is that people don't implement it correctly.  
… It's often destroyed or mangled as it goes through HTTP servers.  
… We can only pick one filename suffix.  
… We can't use profile to do that.  

> *Orie Steele:* feels like implementing profile correctly, is easier than using multiple suffixes consistently.

> *Dmitri Zagidulin:* `@orie` - agree.

> *Ted Thibodeau Jr.:* *sighs* people who don't conform to specs (e.g., by misconfiguring servers or otherwise breaking the `profile` option) should not be justification for not conforming to specs!

> *Manu Sporny:* It's not easier to implement profile correctly, Orie. :) -- We have multiple decades of experience now that people don't implement it correctly.

> *Dave Longley:* as long as multiple suffixes follows the rule that each subtype fits within the same wider syntax, things are kept clean (and correct) ... i think trouble has come from people thinking it's for enveloping, when it isn't.

> *Manu Sporny:* ^ yes, that.

> *Manu Sporny:* (to what dlongley said).

**Dmitri Zagidulin:** I want to agree with Orie that hierarchical multiple suffixes is way more trouble than it's worth.  
… I don't think there's any need for hierarchical multiple suffixes.  

> *Orie Steele:* +1 ivan.

> *Dmitri Zagidulin:* +1 to at risk.

**Ivan Herman:** If we want to keep multiple suffixes, we should put the whole concept as being at risk in the document.  
… Put it as at risk.  
… We can see what happens at the IETF.  
… If it goes through at IETF, we can remove the at risk.  
… Otherwise we can reconsider.  
… We have to defend ourselves.  

**Brent Zundel:** We're relying on multiple suffixes but not normatively pointing to them.  

> *Manu Sporny:* Note that JSON-LD has gone through multiple RECs w/ this in there: [https://w3c.github.io/json-ld-syntax/#structured-extension-ld-json](https://w3c.github.io/json-ld-syntax/#structured-extension-ld-json).

**Ivan Herman:** On the other hand, we are supposed to register media types we use during CR and we can't do it now.  
… We can't ignore this.  
… That's why we have to mark it as at-risk.  

**Orie Steele:** I spoke to our W3C liaison at the IETF - Martin.  
… It might be fine to go into CR that way.  
… To become a recommendation, we have to finish the IETF work.  
… It's OK to go into CR with the references are they are.  
… If we have to correct this, it would be good to be able to do this without going through CR.  
… If marking it as at-risk helps, let's do that.  
… I don't know how the DID spec got to Rec with multiple suffixes.  
… If this isn't resolved before REC, I will formally object.  

> *Orie Steele:* thats not true manu.

> *Orie Steele:* I added the +ld+json structured suffix.

**Manu Sporny:** The DID and JSON-LD specs have gone through REC with multiple structured suffixes.  
… You can treat it as an opaque thing.  
… We can change it to application/vc if we need to.  
… But that's doing the wrong thing.  
… We should fix the IETF's broken suffixes.  
… We should stop routing around this damage at the IETF.  

_See github pull request [json-ld-syntax#415](https://github.com/w3c/json-ld-syntax/pull/415)._




> *Brent Zundel:* [https://www.w3.org/TR/did-core/#application-did-ld-json](https://www.w3.org/TR/did-core/#application-did-ld-json).

> *Orie Steele:* seems like the W3C is responsible for the broken behavior, and I agree that it needs to be fixed.

> *Orie Steele:* I am trying to fix it FWIW.

**Brent Zundel:** The DID spec said that it will be registered as soon as the IETF spec is done.  

**Ivan Herman:** The only thing I don't want is to go there and not even mention the problem.  
… If we include a note like we did in the DID spec, that may be enough.  

**Brent Zundel:** What are the steps forward for the profile parameter issue?  
… We could prohibit their use or say how to use them.  

**Manu Sporny:** We tell people how to use profile parameters correctly.  
… Expect people to break this and it not to work correctly across the ecosystem.  
… We should warn them that it has never worked out well when people rely on the profile parameter.  

> *Ivan Herman:* +1 to manu's direction.

**Brent Zundel:** We don't need to include doomsaying.  
… We would need someone willing to be assigned to the issue.  
… If you feel strongly that doing nothing is the right approach, then please volunteer to be assigned.  
… If noone is assigned, then by default, we will be doing nothing about this.  

> *Orie Steele:* the media types impact the "verification algorithm".

**Ivan Herman:** All these options we are talking about are editorial?  

> *Orie Steele:* are they still editorial?

**Brent Zundel:** The media types affect the verification algorithm.  
… We don't have anyone willing to be assigned.  

> *Orie Steele:* the reason I am not volunteering is that we are removing the dependency from vc-jose-cose.

</details>

**Comment by msporny** [comment]
I agree with @selfissued, [more here](https://github.com/w3c/vc-data-model/issues/1364#issuecomment-1837588721).

**Comment by OR13** [comment]
I prefer the validation / verification algorithm to explicitly prohibit parameters before this is closed.

https://github.com/w3c/vc-data-model/pull/1338#discussion_r1411374594

**Comment by msporny** [comment]
PR #1382 has been raised to address this issue. This issue will be closed once PR #1382 has been merged.

**Comment by msporny** [comment]
PR #1382 has been merged, closing.
